modules/tremorBidAdapter.js

```diff
@@ -0,0 +1,169 @@
+    // get the sizes
+    let width, height;
+    if (utils.isArray(bid.sizes) && bid.sizes.length === 2 && (!isNaN(bid.sizes[0]) && !isNaN(bid.sizes[1]))) {
+      width = bid.sizes[0];
+      height = bid.sizes[1];
+    } else if (typeof bid.sizes === 'object') {
+      // take the primary (first) size from the array
+      width = bid.sizes[0][0];
+      height = bid.sizes[0][1];
+    }
+    if (width && height) {
+      let scheme = ((document.location.protocol === 'https:') ? 'https' : 'http') + '://';
+      let url = scheme + bid.params.supplyCode + ENDPOINT + '?adCode=' + bid.params.adCode;
+
+      url += ('&playerWidth=' + width);
+      url += ('&playerHeight=' + height);
+      url += ('&srcPageUrl=' + encodeURIComponent(document.location.href));
+
+      OPTIONAL_PARAMS.forEach(param => {
+        if (bid.params[param]) {
+          url += ('&' + param + '=' + bid.params[param]);
+        }
+      });
+
+      url = (url + '&fmt=json');
+
+      return url;
+    }
+  }
+
+  /* Notify Prebid of bid responses so bids can get in the auction */
+  function handleResponse(bidReq, response) {
+    let bidResult;
+
+    try {
+      bidResult = JSON.parse(response);
+    } catch (error) {
+      utils.logError(error);
+    }
+
+    if (!bidResult || bidResult.error) {
+      let errorMessage = `in response for ${baseAdapter.getBidderCode()} adapter`;
+      if (bidResult && bidResult.error) {
+        errorMessage += `: ${bidResult.error}`;
+      }
+      utils.logError(errorMessage);
+
+      // signal this response is complete
+      bidmanager.addBidResponse(bidReq.placementCode, createBid(STATUS.NO_BID));
+    }
+
+    if (bidResult.seatbid && bidResult.seatbid.length > 0) {
+      bidResult.seatbid[0].bid.forEach(tag => {
+        let status = STATUS.GOOD;
+        const bid = createBid(status, bidReq, tag);
+        bidmanager.addBidResponse(bidReq.placementCode, bid);
+      });
+    } else {
+      // signal this response is complete with no bid
+      bidmanager.addBidResponse(bidReq.placementCode, createBid(STATUS.NO_BID));
+    }
+  }
+
+  /**
+   * We require the ad code and the supply code to generate a tag url
+   * @param bid
+   * @returns {*}
+   */
+  function valid(bid) {
+    if (bid.params.adCode && bid.params.supplyCode) {
+      return bid;
+    } else {
+      utils.logError('missing bid params');
+    }
+  }
+
+  /**
+   * Create and return a bid object based on status and tag
+   * @param status
+   * @param reqBid
+   * @param response
+   */
+  function createBid(status, reqBid, response) {
+    let bid = bidfactory.createBid(status, reqBid);
+    bid.code = baseAdapter.getBidderCode();
+    bid.bidderCode = baseAdapter.getBidderCode();
+
+    if (response) {
+      bid.cpm = response.price;
+      bid.crid = response.crid;
+      bid.vastXml = response.adm;
+      // This is a dummy vast URL, Tremor sends back the vast XML (the line above).
+      bid.vastUrl = 'http://cdn.tremorhub.com/static/dummy.xml';
```

Comment:
Bid responses with a `vastXml` property will now pass validation with #1556 in master, so if Tremor sends xml and not a vast url, this line can be dropped